cpp/benchmarks/join/generate_input_tables.cuh

```diff
@@ -0,0 +1,283 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+#ifndef __GENERATE_INPUT_TABLES_CUH
+#define __GENERATE_INPUT_TABLES_CUH
+
+#include <curand.h>
+#include <curand_kernel.h>
+#include <thrust/sequence.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/distance.h>
+#include <cassert>
+
+#include <cudf/cudf.h>
+#include <rmm/rmm.h>
+#include <utilities/error_utils.hpp>
+#include <utilities/device_atomics.cuh>
+
+
+__global__ static void init_curand(curandState * state, const int nstates)
+{
+    int ithread = threadIdx.x + blockIdx.x * blockDim.x;
+
+    if (ithread < nstates) {
+        curand_init(1234ULL, ithread, 0, state + ithread);
+    }
+}
+
+
+template<typename key_type, typename size_type>
+__global__ static void init_build_tbl(
+    key_type* const build_tbl, const size_type build_tbl_size,
+    const key_type rand_max,
+    const bool uniq_build_tbl_keys,
+    key_type* const lottery, const size_type lottery_size,
+    curandState * state, const int num_states)
+{
+    static_assert(std::is_signed<key_type>::value, "key_type needs to be signed for lottery to work");
+
+    const int start_idx = blockIdx.x * blockDim.x + threadIdx.x;
+    const key_type stride = blockDim.x * gridDim.x;
+    assert(start_idx < num_states);
+
+    curandState localState = state[start_idx];
+
+    for (size_type idx = start_idx; idx < build_tbl_size; idx += stride) {
+        const double x = curand_uniform_double(&localState);
+
+        if (uniq_build_tbl_keys) {
+	        // If the build table keys need to be unique, go through lottery array from lottery_idx until finding a key
+            // which has not been used (-1). Mark the key as been used by atomically setting the spot to -1.
+
+            size_type lottery_idx = x * lottery_size;
+            key_type lottery_val = -1;
+
+            while (-1 == lottery_val) {
+                lottery_val = lottery[lottery_idx];
+
+                if (-1 != lottery_val) {
+                    lottery_val = atomicCAS<key_type>(lottery + lottery_idx, lottery_val, -1);
+                }
+
+                lottery_idx = (lottery_idx + 1) % lottery_size;
+            }
+
+            build_tbl[idx] = lottery_val;
+        } else {
+            build_tbl[idx] = x * rand_max;
+        }
+    }
+
+    state[start_idx] = localState;
+}
+
+
+template<typename key_type, typename size_type>
+__global__ void init_probe_tbl(
```

Comment:
I feel like benchmarks should be as simple as possible. You shouldn't need new CUDA kernels that aren't already part of the library.

Reply:
This is some special logic for guaranteeing statistical properties of the input tables for the Join, i.e., the matching rate between the two tables. I think a custom kernel is justified here.